addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/internal/ClientSideHandler.java

```diff
@@ -64,81 +78,210 @@ public void handleMessage(HttpMessageHandlerContext ctx, HttpMessage msg) {
+            SessionManagementRequestDetails smReqDetails =
+                    AuthUtils.findSessionTokenSource(st.getValue(), firstHrefId);
+            if (smReqDetails != null) {
+                candidate = new AuthRequestDetails(msg);
+                if (candidate.isBetterThan(authReq, headerConfigs)) {
+                    LOGGER.debug(
+                            "Found better auth candidate {} {}",
+                            msg.getRequestHeader().getMethod(),
+                            msg.getRequestHeader().getURI());
+                    authReq = candidate;
                 }
             }
-
-            if (authMsg != null && View.isInitialised()) {
-                AuthUtils.logUserMessage(
-                        Level.INFO,
-                        Constant.messages.getString(
-                                "authhelper.auth.method.browser.output.sessionid", getHrefId(msg)));
-            }
-        }
-        if (firstHrefId == 0 && msg.getHistoryRef() != null) {
-            firstHrefId = msg.getHistoryRef().getHistoryId();
         }
     }
 
-    private String getHrefId(HttpMessage msg) {
-        if (msg.getHistoryRef() != null) {
-            return Integer.toString(msg.getHistoryRef().getHistoryId());
-        }
-        return "?";
-    }
-
     public HttpMessage getAuthMsg() {
-        return authMsg;
+        return authReq.getMsg();
     }
 
     public void resetAuthMsg() {
-        this.authMsg = null;
+        this.authReq = null;
+    }
+
+    protected static boolean isBetterThan(
+            SessionManagementRequestDetails smrd1, SessionManagementRequestDetails smrd2) {
+        if (smrd2 == null) {
+            return true;
+        }
+        if (smrd1.getConfidence() > smrd2.getConfidence()) {
+            return true;
+        }
+        if (smrd1.getConfidence() < smrd2.getConfidence()) {
+            return false;
+        }
+        return smrd1.getTokens().size() > smrd2.getTokens().size();
     }
 
     public HttpMessage getFallbackMsg() {
         return fallbackMsg;
     }
+
+    protected static List<Pair<String, String>> extractKeyValuePairs(String input) {
+        List<Pair<String, String>> keyValuePairs = new ArrayList<>();
+        Pattern pattern = Pattern.compile("\\{%([^:]+):([^%]+)%}");
+        Matcher matcher = pattern.matcher(input);
+
+        while (matcher.find()) {
+            keyValuePairs.add(new Pair<>(matcher.group(1), matcher.group(2)));
+        }
+
+        return keyValuePairs;
+    }
+
+    protected static int messageTokenCount(HttpMessage msg, List<Pair<String, String>> kvPairs) {
+        int count = 0;
+        Collection<SessionToken> tokens = AuthUtils.getAllTokens(msg, false).values();
+
+        for (Pair<String, String> kvPair : kvPairs) {
+            for (SessionToken token : tokens) {
+                if (token.getSource().equals(kvPair.first)
+                        && token.getKey().equals(kvPair.second)) {
+                    count += 1;
```

Comment:
Why not `count++` ? Not that this is wrong, it just seems strange for our code base.